src/hardware/sblaster.cpp

```diff
@@ -769,59 +766,68 @@ static void PlayDMATransfer(uint32_t bytes_requested)
 			sb.adpcm.haveref=false;
 			sb.adpcm.reference=sb.dma.buf.b8[0];
 			sb.adpcm.stepsize=MIN_ADAPTIVE_STEP_SIZE;
-			i++;
+			++i;
 		}
 		assert(samples == 0);
 		while (i < bytes_read) {
-			MixTemp[samples++] = decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 6) & 0x3);
-			MixTemp[samples++] = decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 4) & 0x3);
-			MixTemp[samples++] = decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 2) & 0x3);
-			MixTemp[samples++] = decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 0) & 0x3);
-
-			frames = check_cast<uint16_t>(samples / channels);
-			sb.chan->AddSamples_m8(frames, maybe_silence(samples, MixTemp));
-			samples = 0;
+ 			const uint8_t decoded[] = {
+			        decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 6) & 0x3),
+			        decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 4) & 0x3),
+			        decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 2) & 0x3),
+			        decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 0) & 0x3)};
+
+			constexpr uint8_t num_decoded = ARRAY_LEN(decoded);
+			sb.chan->AddSamples_m8(num_decoded, maybe_silence(num_decoded, decoded));
+			samples += num_decoded;
 			++i;
 		}
+		frames = check_cast<uint16_t>(samples / channels);
+
 		break;
 	case DSP_DMA_3:
 		bytes_read = ReadDMA8(bytes_to_read);
 		if (bytes_read && sb.adpcm.haveref) {
 			sb.adpcm.haveref=false;
 			sb.adpcm.reference=sb.dma.buf.b8[0];
 			sb.adpcm.stepsize=MIN_ADAPTIVE_STEP_SIZE;
-			i++;
+			++i;
 		}
 		assert(samples == 0);
 		while (i < bytes_read) {
-			MixTemp[samples++] = decode_ADPCM_3_sample((sb.dma.buf.b8[i] >> 5) & 0x7);
-			MixTemp[samples++] = decode_ADPCM_3_sample((sb.dma.buf.b8[i] >> 2) & 0x7);
-			MixTemp[samples++] = decode_ADPCM_3_sample((sb.dma.buf.b8[i] & 0x3) << 1);
-
-			frames = check_cast<uint16_t>(samples / channels);
-			sb.chan->AddSamples_m8(frames, maybe_silence(samples, MixTemp));
-			samples = 0;
+			const uint8_t decoded[] = {
+			        decode_ADPCM_3_sample((sb.dma.buf.b8[i] >> 5) & 0x7),
+			        decode_ADPCM_3_sample((sb.dma.buf.b8[i] >> 2) & 0x7),
+			        decode_ADPCM_3_sample((sb.dma.buf.b8[i] & 0x3) << 1)};
```

Comment:
"Implicit conversion loses integer precision: 'int' to 'uint8_t' (aka 'unsigned char')"

Would that pacify compiler and not break everything? `decode_ADPCM_3_sample((sb.dma.buf.b8[i] << 1) & 0b110)};`

Reply:
Rewrote it to move all the ADPCM decoded inside functions (and de-duped all the code). Looks like it's passing now.